poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+
+// These are messages used for signatures
+var sigMsg []byte = []byte("Promise Signature")
+var sigBlameMsg []byte = []byte("Promise Blame Signature")
+
+/* Promise structs are mechanisms by which a server can promise other servers
+ * that an abstract.Secret will be availble even if the secret's owner goes
+ * down. The secret to be promised will be sharded into shared secrets that can
+ * be combined using Lagrange Interpolation to produce the original secret.
+ * Other servers will act as insurers maintaining a share. If a client ever
+ * needs the secret to be reconstructed, it can contact the insurers to regain
+ * the shares and reconstruct the secret.
+ *
+ * Note to Developers:
+ *       The insurers and secrets arrays should remain synchronized. In other
+ *       words, insurers[i] and secrets[i] should both refer to the same server.
+ *
+ * Note to users of this code:
+ *
+ *   Here is a list of methods that should be called by each type of server:
+ *
+ * - Promisers
+ *   * ConstructPromise
+ *
+ * - Insurers
+ *   * VerifyPromise
+ *   * VerifyShare
+ *   * Sign
+ *   * RevealShare
+ *   * Blame
+ *
+ * - Clients
+ *   * VerifyPromise
+ *   * VerifyRevealedShare
+ *
+ * - All
+ *   * UnmarshalInit
+ *   * VerifySignature
+ *   * VerifyBlame
+ */
+type Promise struct {
+
+	// The id of the promise used to differentiate it from others
+	// The id is the short term public key of the private key being promised
+	id string
+
+	// The cryptographic suite to use for the shared secrets
+	shareSuite abstract.Suite
+
+	// The suite used for the long term keys stored in the Promise
+	keySuite abstract.Suite
+
+	// The minimum number of shares needed to reconstruct the secret
+	t int
+
+	// The minimum number of PromiseSignatures approving the Promise that
+	// are needed before the Promise is certified. t <= r <= n
+	r int
+
+	// The total number of shares
+	n int
+
+	// The long-term public key of the promiser
+	pubKey abstract.Point
+
+	// The public polynomial that is used to verify the shared secrets
+	pubPoly poly.PubPoly
+
+	// A list of servers who will act as insurers of the Promise. The list
+	// contains the long-term public keys of the insurers
+	insurers []abstract.Point
+
+	// The list of shared secrets to be sent to the insurers. They are
+	// encrypted with Diffie-Hellman shared secrets between the insurer
+	// and the promiser.
+	secrets []abstract.Secret
+}
+
+/* Constructs a new Promise to guarentee a secret.
+ *
+ * Arguments
+ *    secretPair   = the keypair of the secret to be promised
+ *    longPair     = the long term keypair of the promiser
+ *    t            = minimum number of shares needed to reconstruct the secret.
+ *    r            = minimum PromiseSignatures needed to certify the Promise
+ *    insurers     = a list of the long-term public keys of the insurers.
+ *
+ * Returns
+ *   A newly constructed Promise
+ */
+func (p *Promise) ConstructPromise(secretPair *config.KeyPair,
```

Comment:
I think there should be only one constructor/init function, such as this, which is called both on the promiser node and on other nodes that will receive and unmarshal a promise.  All other nodes should also call this Init function first with valid suite, t, r, and insurers; let's call those fields "configuration information" that all nodes should know and agree on a priori.  Really the *only* difference in initialization here should be that the promiser includes the secret to be secret-shared, whereas on other nodes that secret can be 'nil' (and in that case this function will skip the secret-sharing part).